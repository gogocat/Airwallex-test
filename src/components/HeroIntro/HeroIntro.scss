.hero-intro {
    display: flex;
    height: 100%;
    align-items: center;
}

.dialog-close {
    padding: 8px;
    text-align: right;
}

.dialog-content {
    padding-top: 0!important;
}