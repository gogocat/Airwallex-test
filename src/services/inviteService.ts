import { inviteAPI } from 'constants/apis';
import { safeJsonStringify } from 'utils';

interface IIviteData {
    name: string,
    email: string
}

export async function postInvite(data: IIviteData) {
    try {
        const response = await fetch(inviteAPI, {
            method: 'POST',
            mode: 'cors',
            headers: {
                'Content-Type': 'application/json',
              },
            body: safeJsonStringify(data),
        });
        
        if (response.status !== 200) {
            throw new Error('Registration error');
        }
    
        return response.json();
    } catch(err) {
        throw err;
    }
}
