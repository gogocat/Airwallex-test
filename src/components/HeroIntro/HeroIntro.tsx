import React from 'react';
import Button from '@mui/material/Button';
import Stack from '@mui/material/Stack';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import Container from '@mui/material/Container';
import Dialog from '@mui/material/Dialog';
import IconButton from '@mui/material/IconButton';
import CloseIcon from '@mui/icons-material/Close';
import DialogContent from '@mui/material/DialogContent';
import useMediaQuery from '@mui/material/useMediaQuery';
import { useTheme } from '@mui/material/styles';
import InviteForm from 'components/Forms/InviteForm';

import './HeroIntro.scss';

function HeroIntro() {
    const [open, setOpen] = React.useState(false);
    const theme = useTheme();
    const fullScreen = useMediaQuery(theme.breakpoints.down('md'));

    const handleDialogOpen = () => {
        setOpen(true);
    };

    const handleDialogClose = (event: any, reason: string) => {
        if (reason === 'backdropClick') {
            return;
        }
        setOpen(false);
    };

    return (
        <Box
            sx={{
                bgcolor: 'background.paper',
                pt: 8,
                pb: 6,
            }}
            className="hero-intro"
        >
            <Container maxWidth="sm">
                <Typography
                    component="h1"
                    variant="h2"
                    align="center"
                    color="text.primary"
                    gutterBottom
                >
                    A better way to enjoy everyday.
                </Typography>
                <Typography variant="h5" align="center" color="text.secondary" paragraph>
                    Be the first to know when we launch.
                </Typography>
                <Stack
                    sx={{ pt: 4 }}
                    direction="row"
                    spacing={2}
                    justifyContent="center"
                >
                    <Button 
                        variant="contained"
                        onClick={handleDialogOpen}
                    >
                        Request an invite
                    </Button>
                </Stack>
            </Container>

            <Dialog
                fullScreen={fullScreen}
                open={open}
                onClose={handleDialogClose}
                aria-labelledby="responsive-dialog-title"
            >
                <div className="dialog-close">
                    <IconButton onClick={(e)=>{handleDialogClose(e, '')}}>
                        <CloseIcon />
                    </IconButton>
                </div>
                <DialogContent className="dialog-content">
                    <InviteForm
                        onSuccess={handleDialogClose}
                    />
                </DialogContent>
            </Dialog>
        </Box>
    );
}

export default HeroIntro;
