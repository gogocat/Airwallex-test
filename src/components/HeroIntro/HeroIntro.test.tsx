import React from 'react';
import {
    render,
    screen,
    act,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import HeroIntro from './HeroIntro';
import { postInvite } from 'services/inviteService';

jest.mock('services/inviteService', () => ({ 
    postInvite: jest.fn(),
}));

describe('HeroIntro', () => {
    it('should renders with HeroIntro content', () => {
        const {
            asFragment,
        } = render(<HeroIntro />);
    
        const introTitle = screen.getByText('A better way to enjoy everyday.');

        expect(introTitle).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should trigger invite dialog', async () => {
        const user = userEvent.setup();

        render(<HeroIntro />);
    
        const inivteButton = screen.getByRole('button', { name: 'Request an invite' });

        user.click(inivteButton);

        const dialogHeading = await screen.findByText('Request an invite', { selector: 'h2'});

        expect(dialogHeading).toBeInTheDocument();
    });

    it('should submit invite form', async () => {
        const user = userEvent.setup();

        const mockData = {
            fullName: 'Adam',
            email: 'test@test.com'
        };

        render(<HeroIntro />);
    
        const inivteButton = screen.getByRole('button', { name: 'Request an invite' });

        user.click(inivteButton);
        // wait for dialog shows up 
        await screen.findByText('Request an invite', { selector: 'h2'});

        const fullNameInput = screen.getByLabelText('Full name');
        const emailInput = screen.getByLabelText('Email');
        const confirmEmailInput = screen.getByLabelText('Confirm email');
        const sendButton = screen.getByRole('button', { name: 'Send' });

        await act(async () => {
            await user.type(fullNameInput, mockData.fullName);
            await user.type(emailInput, mockData.email);
            await user.type(confirmEmailInput, mockData.email);
        });

        expect(fullNameInput).toHaveValue(mockData.fullName);
        expect(emailInput).toHaveValue(mockData.email);
        expect(confirmEmailInput).toHaveValue(mockData.email);

        await act(async () => {
            await user.click(sendButton);
        });

        // const sumittingButton = screen.getByRole('button', { name: 'Sending, please wait...' });

        expect(postInvite).toHaveBeenCalledTimes(1);
        expect(postInvite).toHaveBeenCalledWith({
            email: mockData.email,
            name: mockData.fullName
        });

        // const successTitle = await screen.findByText('All done!', { selector: 'h2' });
        // expect(sumittingButton).toBeInTheDocument();
    });
});