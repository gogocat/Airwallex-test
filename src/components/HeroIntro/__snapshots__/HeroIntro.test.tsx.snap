// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroIntro should renders with HeroIntro content 1`] = `
<DocumentFragment>
  <div
    class="hero-intro MuiBox-root css-10m0cap"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h2 MuiTypography-alignCenter MuiTypography-gutterBottom css-10759ka-MuiTypography-root"
      >
        A better way to enjoy everyday.
      </h1>
      <p
        class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter MuiTypography-paragraph css-17rfbpd-MuiTypography-root"
      >
        Be the first to know when we launch.
      </p>
      <div
        class="MuiStack-root css-kof1v3-MuiStack-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Request an invite
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
